### PR TITLE
Update trace_link.py to merge Kineto trace with Pytorch execution trace. 

### PR DESCRIPTION
**Pull request by JoongunPark**
This update modifies trace_link.py to incorporate GPU information from the Kineto trace into the PyTorch Execution trace. PyTorch's native execution trace offers insights into CPU operations but lacks GPU details. This enhancement ensures that both CPU and GPU information is captured, providing a more comprehensive view of execution for improved debugging and optimization.

### CODE REVIEW

**Comment by TaekyungHeo** [comment]
@JoongunPark , Could you please provide commands to run trace_link.py with the provided datasets?

**Comment by briancoutinho** [comment]
@JoongunPark  I understand that you want to connect the PyTorch profiler information with the Execution Trace.
These features to parse and analyze traces are actually availabe in HTA project, this was the reason i integrated Execution Trace into HTA, see here
- https://github.com/facebookresearch/HolisticTraceAnalysis/blob/main/examples/execution_trace_demo.ipynb 

These changes are fine but it would be good if we could use HTA for trace analysis. I don't want to prescribe since this is something you are working on but we do end up in duplicate code over time.

**Comment by JoongunPark** [comment]
@briancoutinho, thank you for your input! I want to clarify that our modifications are specifically aimed at datasets collected over different time intervals from Kineto and Pytorch, a scenario that the current HTA does not appear to support as of now. 
We are open to considering HTA for future datasets, but our current needs necessitate these specific features.

File: train/compute/python/tools/trace_link.py [hunk 1]
```diff
@@ -80,8 +82,127 @@ def find_closest_segment(segs, target_length):
...
+            + "and hence it is discarded. gpu_op name: "
+            + str(kineto_gpu_op["name"])
+            + ", ts: "
+            + str(get_timestamp_field(kineto_gpu_op))
+            + ", external_id: "
+            + str(kineto_gpu_op["args"]["External id"])
+        )
+
+    # TODO: In some cases, condition 2 is not met. Seems to be a problem in input from older Kineto.
+    # assert parent_cpu_op and get_timestamp_field(parent_cpu_op) + get_duration_field(parent_cpu_op) > get_timestamp_field(kineto_gpu_op)
```
Comment (TaekyungHeo) [inline train/compute/python/tools/trace_link.py:198]: @JoongunPark Could you please remove commented-out codes?

**Comment by srinivas212** [comment]
> @JoongunPark I understand that you want to connect the PyTorch profiler information with the Execution Trace. These features to parse and analyze traces are actually availabe in HTA project, this was the reason i integrated Execution Trace into HTA, see here
> 
> * https://github.com/facebookresearch/HolisticTraceAnalysis/blob/main/examples/execution_trace_demo.ipynb
> 
> These changes are fine but it would be good if we could use HTA for trace analysis. I don't want to prescribe since this is something you are working on but we do end up in duplicate code over time.

@briancoutinho - valid feedback. Like @JoongunPark already mentioned, HTA does not support this feature today. 

We can and should align on HTA as the single solution for such analysis etc. Two things that could help:  
1. Visibility - it would be great if you can help share the plans and roadmap etc. 
2. Figure out a path to merging this feature into HTA and enable future development in HTA directly 

Happy to get on a call to discuss this further. 

cc: @louisfeng

File: train/compute/python/tools/trace_link.py (hunk 1 above)
Comment (JoongunPark) [reply train/compute/python/tools/trace_link.py]: @TaekyungHeo Yes. I have removed.

**Comment by JoongunPark** [comment]
> @JoongunPark , Could you please provide commands to run trace_link.py with the provided datasets?

# First, navigate to the param submodule where you'll find sample Pytorch-ET and Kineto input files and that handy merger script
$ cd extern/graph_frontend/param/train/compute/python/
$ pip3 install -r requirements.txt
$ pip3 install .

# Second, unzip the input files 
$ cd ./test/data
$ tar -xzf dlrm_kineto.tar.gz
$ tar -xzf dlrm_pytorch_et.tar.gz
$ cd ../../

# Now, we'll fire up the script that leverages the already whipped up Pytorch-ET+Kineto traces (from a single training loop of DLRM on 8 GPUs) and fuses them into that enhanced ET we keep talking about
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_0.json --kineto-file test/data/dlrm_kineto/worker0* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_1.json --kineto-file test/data/dlrm_kineto/worker1* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_2.json --kineto-file test/data/dlrm_kineto/worker2* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_3.json --kineto-file test/data/dlrm_kineto/worker3* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_4.json --kineto-file test/data/dlrm_kineto/worker4* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_5.json --kineto-file test/data/dlrm_kineto/worker5* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_6.json --kineto-file test/data/dlrm_kineto/worker6* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_7.json --kineto-file test/data/dlrm_kineto/worker7* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.__next__'

**Comment by TaekyungHeo** [comment]
@JoongunPark , The following command sequence works. Please provide a full command sequence next time.
```
$ git clone --recurse-submodules git@github.com:facebookresearch/param.git
$ cd param
$ conda create -n param python=3.8.0
$ conda activate para
$ pip3 install -r requirements.txt
$ cd train/compute/python/
$ pip3 install .
$ pip3 install -r requirements.txt
$ cd test/data
$ tar -xzf dlrm_kineto.tar.gz
$ tar -xzf dlrm_pytorch_et.tar.gz
$ cd ../../

$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_0.json --kineto-file test/data/dlrm_kineto/worker0* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_1.json --kineto-file test/data/dlrm_kineto/worker1* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_2.json --kineto-file test/data/dlrm_kineto/worker2* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_3.json --kineto-file test/data/dlrm_kineto/worker3* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_4.json --kineto-file test/data/dlrm_kineto/worker4* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_5.json --kineto-file test/data/dlrm_kineto/worker5* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_6.json --kineto-file test/data/dlrm_kineto/worker6* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
$ python3 ./tools/trace_link.py --et-file test/data/dlrm_pytorch_et/dlrm_eg_7.json --kineto-file test/data/dlrm_kineto/worker7* --exact-match --annotation 'enumerate(DataLoader)#_MultiProcessingDataLoaderIter.next'
```

File: train/compute/python/tools/trace_link.py [hunk 2]
```diff
@@ -375,23 +835,36 @@ def approximate_match(kineto_et_events, et_nodes):
...
-
-        et_plus_file = args.et_file.replace(".json", "_plus.json")
-        logger.info(f"Enhanced execution trace dumped to {et_plus_file}")
-
-        write_dictionary_to_json_file(et_plus_file, et)
+        logger.info(
+            "This script works only with the exact match mode,"
+            + "please add the --exact-match flag to your script and run it again"
+        )
+        assert False
```
Comment (TaekyungHeo) [inline train/compute/python/tools/trace_link.py:870]: @JoongunPark  Please replace this with raising an exception.

**Comment by briancoutinho** [comment]
> @briancoutinho, thank you for your input! I want to clarify that our modifications are specifically aimed at datasets collected over different time intervals from Kineto and Pytorch (which are also collected separately), a scenario that the current HTA does not appear to support as of now. We are open to considering HTA for future datasets, but our current needs necessitate these specific features.

Yep this makes sense, long term we do intend to support non overlap case as well in HTA. I was hitting some issues running the trace_link for some large traces (param is a git submodule in HTA so we do re-use your code :) ).
Let's sync up on the future direction in Charka WG

**Comment by srinivas212** [comment]
> > @briancoutinho, thank you for your input! I want to clarify that our modifications are specifically aimed at datasets collected over different time intervals from Kineto and Pytorch (which are also collected separately), a scenario that the current HTA does not appear to support as of now. We are open to considering HTA for future datasets, but our current needs necessitate these specific features.
> 
> Yep this makes sense, long term we do intend to support non overlap case as well in HTA. I was hitting some issues running the trace_link for some large traces (param is a git submodule in HTA so we do re-use your code :) ). Let's sync up on the future direction in Charka WG

@briancoutinho  can you accept this PR if this looks good? This is with the assumption we figure out a path to integration in HTA long term.

**Comment by briancoutinho** [comment]
@srinivas212 sounds good to me:)

**Review by briancoutinho — APPROVED** [review]
Lgtm

File: train/compute/python/tools/trace_link.py (hunk 2 above)
Comment (JoongunPark) [reply train/compute/python/tools/trace_link.py]: @TaekyungHeo Update it as requested.

**Pull request closed without merging**